core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala

```diff
@@ -0,0 +1,201 @@
+package com.mesosphere.usi.core
+
+import java.time.Instant
+
+import com.mesosphere.usi.core.models._
+import com.mesosphere.usi.models._
+
+/**
+  * Container class responsible for keeping track of the frame and cache.
+  *
+  * As a general rule, business logic does not manipulate the USI state directly, but rather does so by returning
+  * effects.
+  */
+private[core] class SchedulerLogic {
+  private var frame: Frame = Frame(Map.empty, Map.empty, Map.empty)
+  private var cachedPendingLaunch = CachedPendingLaunch(Set.empty)
+
+  def processSpecEvent(msg: SpecEvent): FrameEffects = {
+    val (frameWithPodSpecChanges, changedPodIds) = frame.applySpecEvent(msg)
+    frame = frameWithPodSpecChanges
+    val frameEffects = SchedulerLogic.computeNextStateForPods(frameWithPodSpecChanges)(changedPodIds)
+    applyAndUpdateCaches(frameEffects, changedPodIds)
+  }
+
+  /**
+    * Given a frame effects, apply the usi state change effects of the frame effects. Then, update the cache and perform
```

Comment:
I had to read the first sentence three times and I still don't quite understand it 😊 Maybe a diagram would help?

Reply:
It should be simpler now with FrameWithEffects